plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastoreModule.java

```diff
@@ -43,32 +46,53 @@ public void configure(Binder binder)
         newOptionalBinder(binder, HiveMetastoreDecorator.class);
         newExporter(binder).export(HiveMetastore.class)
                 .as(generator -> generator.generatedNameOf(CachingHiveMetastore.class));
+        newSetBinder(binder, Procedure.class).addBinding().toProvider(FlushHiveMetastoreCacheProcedure.class).in(Scopes.SINGLETON);
     }
 
     @Provides
     @Singleton
-    public HiveMetastore createCachingHiveMetastore(
+    public Optional<CachingHiveMetastore> createCachingHiveMetastore(
             NodeManager nodeManager,
             @ForCachingHiveMetastore HiveMetastore delegate,
             CachingHiveMetastoreConfig config,
             CatalogName catalogName,
             Optional<HiveMetastoreDecorator> hiveMetastoreDecorator)
     {
-        HiveMetastore decoratedDelegate = hiveMetastoreDecorator
-                .map(decorator -> decorator.decorate(delegate))
-                .orElse(delegate);
-
-        if (!nodeManager.getCurrentNode().isCoordinator()) {
+        if (!nodeManager.getCurrentNode().isCoordinator() || !isCacheEnabled(config)) {
             // Disable caching on workers, because there currently is no way to invalidate such a cache.
             // Note: while we could skip CachingHiveMetastoreModule altogether on workers, we retain it so that catalog
             // configuration can remain identical for all nodes, making cluster configuration easier.
-            return decoratedDelegate;
+            return Optional.empty();
         }
 
-        Executor executor = new ReentrantBoundedExecutor(
-                newCachedThreadPool(daemonThreadsNamed("hive-metastore-" + catalogName + "-%s")),
-                config.getMaxMetastoreRefreshThreads());
+        return Optional.of(cachingHiveMetastore(
+                decorateHiveMetastore(delegate, hiveMetastoreDecorator),
+                new ReentrantBoundedExecutor(
+                        newCachedThreadPool(daemonThreadsNamed("hive-metastore-" + catalogName + "-%s")),
+                        config.getMaxMetastoreRefreshThreads()),
+                config));
+    }
+
+    private boolean isCacheEnabled(CachingHiveMetastoreConfig config)
+    {
+        return config.getMetastoreCacheTtl().toMillis() != 0 && config.getMetastoreCacheMaximumSize() != 0;
+    }
+
+    @Provides
+    @Singleton
+    public HiveMetastore createHiveMetastore(
+            @ForCachingHiveMetastore HiveMetastore delegate,
+            Optional<HiveMetastoreDecorator> hiveMetastoreDecorator,
+            Optional<CachingHiveMetastore> cachingMetastore)
+    {
+        return cachingMetastore.map(metastore -> (HiveMetastore) metastore)
+                .orElseGet(() -> decorateHiveMetastore(delegate, hiveMetastoreDecorator));
```

Comment:
This is tricky. Document why `decorateHiveMetastore` is applied in one case, but not the other case.
Better yet, move `decorateHiveMetastore` application to a single Guice place.

Reply:
Done https://github.com/trinodb/trino/pull/10251#discussion_r770492727